sycl/source/detail/queue_impl.cpp

```diff
@@ -192,24 +199,20 @@ void queue_impl::wait(const detail::code_location &CodeLoc) {
   TelemetryEvent = instrumentationProlog(CodeLoc, Name, StreamID, IId);
 #endif
 
-  std::vector<std::shared_ptr<event_impl>> Events;
+  vector_class<std::weak_ptr<event_impl>> Events;
+  vector_class<event> USMEvents;
   {
-    std::lock_guard<mutex_class> Guard(MMutex);
-    for (std::weak_ptr<event_impl> &EventImplWeakPtr : MEvents)
-      if (std::shared_ptr<event_impl> EventImplPtr = EventImplWeakPtr.lock())
-        Events.push_back(EventImplPtr);
+    std::lock_guard<mutex_class> Lock(MMutex);
+    Events.swap(MEvents);
+    USMEvents.swap(MUSMEvents);
```

Comment:
Why `swap`? There is still a need to have the list of events after it's been waited for, isn't it?
Also, if `swap` is essential you may want to employ move-assignment. I believe it'll be more efficient than swap.

Reply:
> There is still a need to have the list of events after it's been waited for, isn't it?
AFAIK there are not needed anymore. Could you please clarify when they are needed?
Agree with std::move() instead of std::swap, will change.

Reply:
> AFAIK there are not needed anymore. Could you please clarify when they are needed?

OK, then. It was more like a question if the user needs the events already been waited for.